src/R8R.sol

```diff
@@ -0,0 +1,170 @@
+    // === STORAGE VARIABLES ===
+    // =========================
+
+    address public robot;
+    uint256 public gameId;
+    uint256 public gameEntryPriceInEth;
+
+    mapping(uint256 gameId => Game) games;
+
+    IERC20[] public gameTokensAllowedList; // list of ERC20 tokens users can use to enter the game
+    uint256[] public gameTokensEntryPriceList; // list of entry price for the ERC20 tokens
+
+    struct Game {
+        uint256 gameId;
+        address[] players;
+        uint256[] playerRatings;
+        uint256 robotRating;
+        address winner;
+        uint256 winnerPayout;
+    }
+
+    // ==============
+    // === EVENTS ===
+    // ==============
+
+    event GameCreated(uint256 gameId);
+    event PlayerJoinedGameWithEth(address player, uint256 gameId);
+    event PlayerJoinedGameWithTokens(address player, uint256 gameId);
+    event GameEnded();
+
+    // ==============
+    // === ERRORS ===
+    // ==============
+    error Error__TokenTransferFailed();
+
+    // ===================
+    // === CONSTRUCTOR ===
+    // ===================
+
+    constructor(
+        address _robot,
+        uint256 _gameEntryPriceInEth,
+        IERC20[] memory _gameTokensAllowedList,
+        uint256[] memory _gameTokensEntryPriceList
+    ) Ownable(msg.sender) {
+        robot = _robot;
+        gameEntryPriceInEth = _gameEntryPriceInEth;
+        gameTokensAllowedList = _gameTokensAllowedList;
+        gameTokensEntryPriceList = _gameTokensEntryPriceList;
+    }
+
+    // =========================
+    // === PUBLIC FUNCTIONS ====
+    // =========================
+
+    // @notice called by AI wallet to initialise a new game
+    // @notice creates a new Game struct that players + ratings can then be added to when they join
+    function createGame() public onlyRobot {
+        // increment gameId
+        gameId = gameId + 1;
+
+        // empty placeholder arrays for players + their ratings (until someone joins the game)
+        address[] memory playersArray;
+        uint256[] memory playersRatingsArray;
+
+        // create game
+        Game memory game = Game({
+            gameId: gameId,
+            players: playersArray,
+            playerRatings: playersRatingsArray,
+            robotRating: 0,
+            winner: address(0), // set this to our treasury address upon initial game creation??
+            winnerPayout: 0
+        });
+
+        // add new game to game mappings indexed by gameId
+        games[gameId] = game;
+
+        // emit event
+        emit GameCreated(gameId);
+    }
+
+    // @notice allows new players to join a selected game
+    // @param user selects a _gameId to play & submits a rating between 1 - 100 for the image
+    function joinGameWithEth(uint256 _gameId, uint256 _playerRating) public payable nonReentrant {
+        // Checks
+        require(msg.value > 0, "Please send Eth to enter");
+        require(msg.value == gameEntryPriceInEth, "Please send correct amount of Eth to enter");
```

Comment:
Let's do msg.value >= gameEntryPriceInEth and have some refund logic in case they send to much. This way the tx won't fail.